### PR TITLE
Misc fixes to button, add smoothing to simulatedhand

### PR DESCRIPTION
**Pull request by AsoboStudio**
Couple of fixes on MRTK:
* Pressable buttons was not working any more if the button was disabled in the pressed Event
* Add Lerp in SimulatedHand not to miss any frame in the editor
* Added OpenSky pose

### CODE REVIEW

**Comment by wiwei** [comment]
@AsoboStudio thanks for contributing!

Notes on the branch, mrtk_release is where releases get cut to, so we don't put things directly into there - I've updated the branch to be mrtk_development (where our changes are going into). 

There are a couple of conflicts associated with changes to how hand simulation code was updated to support new input simulation recording - can you please resolve the conflicts (we don't have write access to your fork)?

File: README.md [hunk 1]
```diff
@@ -21,7 +21,7 @@ MRTK-Unity provides a set of foundational components and features to accelerate
 
 | Branch | CI Status | Docs Status |
 |---|---|---|
-| `mrtk_development` |[![CI Status](https://dev.azure.com/aipmr/MixedRealityToolkit-Unity-CI/_apis/build/status/public/mrtk_development-CI)](https://dev.azure.com/aipmr/MixedRealityToolkit-Unity-CI/_build/latest?definitionId=1)|[![Docs Status](https://dev.azure.com/aipmr/MixedRealityToolkit-Unity-CI/_apis/build/status/public/mrtk_docs)](https://dev.azure.com/aipmr/MixedRealityToolkit-Unity-CI/_build/latest?definitionId=7)
+| `mrtk_development` |[![CI Status](https://dev.azure.com/aipmr/MixedRealityToolkit-Unity-CI/_apis/build/status/public/mrtk_CI?branchName=mrtk_development)](https://dev.azure.com/aipmr/MixedRealityToolkit-Unity-CI/_build/latest?definitionId=15)|[![Docs Status](https://dev.azure.com/aipmr/MixedRealityToolkit-Unity-CI/_apis/build/status/public/mrtk_docs)](https://dev.azure.com/aipmr/MixedRealityToolkit-Unity-CI/_build/latest?definitionId=7)
```
Comment (julenka) [inline README.md]: Was this change intentional?

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/PressableButtons/PressableButton.cs [hunk 2]
```diff
@@ -130,6 +130,11 @@ private void Start()
             }
         }
 
+        private void OnEnable()
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/PressableButtons/PressableButton.cs]: I believe @thalbern actually fixes this in a very similar way in #4150 so I don't think it's needed any more, but thank you for suggesting it.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/PressableButtons/PressableButton.cs [hunk 3]
```diff
@@ -93,14 +93,14 @@ private set
 
                     if (isTouching)
                     {
-                        TouchBegin.Invoke();
+                        TouchBegin?.Invoke();
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/PressableButtons/PressableButton.cs]: Thanks for this!

**Comment by julenka** [comment]
Hi folks, I'm not sure that your most recent changes are actually applicable for the PR fix you are making, could you please remove the last two changes? those are for eye tracking and seem specific to your application.

**Comment by julenka** [comment]
@LukasTonneMS could you please take a look at this PR?

File: Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHand.cs [hunk 4]
```diff
@@ -117,6 +118,7 @@ public void UpdateState(SimulatedHandData handData)
                 }
                 else
                 {
+                    jointPositions[i] = Vector3.SmoothDamp(jointPoses[handJoint].Position, jointPositions[i], ref jointPositionVelocities[i], 0.15f);
```
Comment (lukastoenneMS) [inline Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHand.cs]: The smoothTime should be configurable, not least because we want to be able to disable it for testing purposes. The value should go into the input simulation profile, with a matching entry in the inspector, like so:

```diff
diff --git a/Assets/MixedRealityToolkit.Services/InputSimulation/Editor/MixedRealityInputSimulationProfileInspector.cs b/Assets/MixedRealityToolkit.Services/InputSimulation/Editor/MixedRealityInputSimulationProfileInspector.cs
index 60001148d..17b62a8f3 100644
--- a/Assets/MixedRealityToolkit.Services/InputSimulation/Editor/MixedRealityInputSimulationProfileInspector.cs
+++ b/Assets/MixedRealityToolkit.Services/InputSimulation/Editor/MixedRealityInputSimulationProfileInspector.cs
@@ -47,6 +47,7 @@ public class MixedRealityInputSimulationProfileInspector : BaseMixedRealityToolk
         private SerializedProperty defaultHandDistance;
         private SerializedProperty handDepthMultiplier;
         private SerializedProperty handJitterAmount;
+        private SerializedProperty handSmoothTime;

         private SerializedProperty yawHandCWKey;
         private SerializedProperty yawHandCCWKey;
@@ -104,6 +105,7 @@ protected override void OnEnable()
             defaultHandDistance = serializedObject.FindProperty("defaultHandDistance");
             handDepthMultiplier = serializedObject.FindProperty("handDepthMultiplier");
             handJitterAmount = serializedObject.FindProperty("handJitterAmount");
+            handSmoothTime = serializedObject.FindProperty("handSmoothTime");

             yawHandCWKey = serializedObject.FindProperty("yawHandCWKey");
             yawHandCCWKey = serializedObject.FindProperty("yawHandCCWKey");
@@ -190,6 +192,7 @@ public override void OnInspectorGUI()
                 EditorGUILayout.PropertyField(defaultHandDistance);
                 EditorGUILayout.PropertyField(handDepthMultiplier);
                 EditorGUILayout.PropertyField(handJitterAmount);
+                EditorGUILayout.PropertyField(handSmoothTime);
                 EditorGUILayout.Space();

                 EditorGUILayout.PropertyField(yawHandCWKey);
diff --git a/Assets/MixedRealityToolkit.Services/InputSimulation/MixedRealityInputSimulationProfile.cs b/Assets/MixedRealityToolkit.Services/InputSimulation/MixedRealityInputSimulationProfile.cs
index 6a4d7564b..b441fb7f1 100644
--- a/Assets/MixedRealityToolkit.Services/InputSimulation/MixedRealityInputSimulationProfile.cs
+++ b/Assets/MixedRealityToolkit.Services/InputSimulation/MixedRealityInputSimulationProfile.cs
@@ -151,6 +151,10 @@ public class MixedRealityInputSimulationProfile : BaseMixedRealityProfile
         [Tooltip("Apply random offset to the hand position")]
         private float handJitterAmount = 0.0f;
         public float HandJitterAmount => handJitterAmount;
+        [SerializeField]
+        [Tooltip("Time it takes hands to reach the target mouse position")]
+        private float handSmoothTime = 0.15f;
+        public float HandSmoothTime => handSmoothTime;

         [Header("Hand Rotation Settings")]
         [SerializeField]
```

Comment (lukastoenneMS) [inline Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHand.cs]: I would like to restrict the smoothing to just the mouse movement, which seems to be it's intended purpose. The gesture animation of the actual joints is already animated smoothly, so the damping just makes it feel too "squishy" IMO and isn't necessary to avoid fast motion.

Doing this in the `SimulatedHandDataProvider` instead has the added advantage that smoothing is not applied when hands are simply animated. This is important for the upcoming input recording feature, which will be used for testing and should reproduce the recorded animation as accurately as possible, without additional smoothing.

```diff
diff --git a/Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHandDataProvider.cs b/Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHandDataProvider.cs
index d348e3dec..53b355145 100644
--- a/Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHandDataProvider.cs
+++ b/Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHandDataProvider.cs
@@ -31,6 +31,7 @@ internal class SimulatedHandState
         public bool IsPinching { get; private set; }

         public Vector3 ScreenPosition;
+        private Vector3 velocity = Vector3.zero;
         // Rotation of the hand
         public Vector3 HandRotateEulerAngles = Vector3.zero;
         // Random offset to simulate tracking inaccuracy
@@ -90,17 +91,21 @@ public void SetViewportPosition(Vector3 point)
             ScreenPosition = CameraCache.Main.ViewportToScreenPoint(point);
         }

-        public void SimulateInput(Vector3 mouseDelta, float noiseAmount, Vector3 rotationDeltaEulerAngles)
+        public void SimulateInput(Vector3 mouseDelta, float noiseAmount, Vector3 rotationDeltaEulerAngles, float smoothTime)
         {
+            Vector3 worldPoint = CameraCache.Main.ScreenToWorldPoint(ScreenPosition);
+            Vector3 newScreenPosition = ScreenPosition;
             // Apply mouse delta x/y in screen space, but depth offset in world space
-            ScreenPosition.x += mouseDelta.x;
-            ScreenPosition.y += mouseDelta.y;
-            Vector3 newWorldPoint = CameraCache.Main.ScreenToWorldPoint(ScreenPosition);
+            newScreenPosition.x += mouseDelta.x;
+            newScreenPosition.y += mouseDelta.y;
+            Vector3 newWorldPoint = CameraCache.Main.ScreenToWorldPoint(newScreenPosition);
             newWorldPoint += CameraCache.Main.transform.forward * mouseDelta.z;
-            ScreenPosition = CameraCache.Main.WorldToScreenPoint(newWorldPoint);

-            HandRotateEulerAngles += rotationDeltaEulerAngles;
+            // Apply damping to the hand movement
+            newWorldPoint = Vector3.SmoothDamp(worldPoint, newWorldPoint, ref velocity, smoothTime);

+            ScreenPosition = CameraCache.Main.WorldToScreenPoint(newWorldPoint);
+            HandRotateEulerAngles += rotationDeltaEulerAngles;
             JitterOffset = UnityEngine.Random.insideUnitSphere * noiseAmount;
         }

@@ -295,7 +300,7 @@ private void SimulateUserInput()

             if (isSimulating)
             {
-                state.SimulateInput(mouseDelta, profile.HandJitterAmount, rotationDeltaEulerAngles);
+                state.SimulateInput(mouseDelta, profile.HandJitterAmount, rotationDeltaEulerAngles, profile.HandSmoothTime);

                 if (isAlwaysVisible)
                 {

```

File: Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHandPose.cs [hunk 5]
```diff
@@ -442,6 +447,36 @@ public string GenerateInitializerCode()
             new Vector3(-0.06393f, 0.08967f, -0.03293f), // PinkyDistalJoint
             new Vector3(-0.06748f, 0.09752f, -0.02542f), // PinkyTip
         });
-    }
 
+        static private SimulatedHandPose HandOpenedSky = new SimulatedHandPose(new Vector3[]
```
Comment (lukastoenneMS) [inline Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHandPose.cs]: Out of curiosity: did you use the HandGestureRecorder tool to record this gesture?

File: Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHandPose.cs [hunk 6]
```diff
@@ -158,6 +158,10 @@ public enum GestureId
             /// Victory sign
             /// </summary>
             Victory,
+            /// <summary>
+            /// Open Sky
```
Comment (lukastoenneMS) [inline Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHandPose.cs]: Better description needed, e.g. "Open hand with palm facing up and fingers spread out".

File: Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHandPose.cs (hunk 5 above)
Comment (AsoboStudio) [reply Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHandPose.cs]: Hi Lukas,
At first, I didn't know the existence of this script, so I created my own MonoBehaviour.
When I found the HandGestureRecorder and tried to attach it to a gameobject, a the pop-up  shows : "The script is an editor script.".
So I created my own GestureRecorder

File: Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHandPose.cs (hunk 6 above)
Comment (AsoboStudio) [reply Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedHandPose.cs]: Absolutly, "Open Sky" means nothing

**Comment by luis-valverde-ms** [comment]
@AsoboStudio , thanks again for the contribution but it seems that some changes are still needed for this to make it into the repository. Are you still working on it?

**Comment by wiwei** [comment]
This change has been open for some time and there have been quite a few changes to this area since it's inception. Looking above it looks like one of the issues has been addressed, and others have had some open comments for a while. 

The changes look to be good overall, just a bit out of date now - thanks for sending this out! Feel free to reopen after resolving the conflicts!
